File: src/main/java/com/paragon/client/ui/console/Console.java
package com.paragon.client.ui.console;

import com.paragon.Paragon;
import com.paragon.api.util.Wrapper;
import com.paragon.api.util.render.RenderUtil;
import com.paragon.api.util.render.TextRenderer;
import com.paragon.client.systems.module.impl.client.Colours;
import net.minecraft.client.gui.GuiTextField;
import net.minecraft.client.gui.ScaledResolution;
import org.lwjgl.input.Keyboard;

import java.awt.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Console implements Wrapper, TextRenderer {

    // Title of window
    private String title;

    // Coordinates and dimensions
    private float width, height;

    // List of lines
    private List<String> lines = new ArrayList<>();

    private GuiTextField guiTextField;

    public Console(String title, float width, float height) {
        this.title = title;
        this.width = width;
        this.height = height;

        ScaledResolution scaledResolution = new ScaledResolution(mc);

        guiTextField = new GuiTextField(0, mc.fontRenderer, (int) ((scaledResolution.getScaledWidth() / 2) - (getWidth() / 2)) + 2, (int) (((scaledResolution.getScaledHeight() / 2) - (getHeight() / 2)) + getHeight() - 11), (int) getWidth(), 11);
    }

    public void init() {
        ScaledResolution scaledResolution = new ScaledResolution(mc);

        guiTextField = new GuiTextField(0, mc.fontRenderer, (int) ((scaledResolution.getScaledWidth() / 2) - (getWidth() / 2)) + 2, (int) (((scaledResolution.getScaledHeight() / 2) - (getHeight() / 2)) + getHeight() - 13), (int) getWidth() - 4, 11);
    }

    public void draw(int mouseX, int mouseY) {
        ScaledResolution scaledResolution = new ScaledResolution(mc);

        // Background
        RenderUtil.drawRect((scaledResolution.getScaledWidth() / 2f) - (getWidth() / 2f), (scaledResolution.getScaledHeight() / 2f) - (getHeight() / 2f), getWidth(), getHeight(), 0x90000000);

        RenderUtil.drawRect((scaledResolution.getScaledWidth() / 2f) - (getWidth() / 2f), (scaledResolution.getScaledHeight() / 2f) - (getHeight() / 2f), getWidth(), 14, new Color(23, 23, 23).getRGB());

        // Title
        renderText(getTitle(), ((scaledResolution.getScaledWidth() / 2f) - (getWidth() / 2f)) + 3, ((scaledResolution.getScaledHeight() / 2f) - (getHeight() / 2f)) + 3, -1);

        // Border
        RenderUtil.drawBorder((scaledResolution.getScaledWidth() / 2f) - (getWidth() / 2f), (scaledResolution.getScaledHeight() / 2f) - (getHeight() / 2f), getWidth(), getHeight(), 1, Colours.mainColour.getValue().getRGB());

        // Separator
        RenderUtil.drawRect((scaledResolution.getScaledWidth() / 2f) - (getWidth() / 2f), (scaledResolution.getScaledHeight() / 2f) - (getHeight() / 2f) + 13, getWidth(), 1, Colours.mainColour.getValue().getRGB());

        // Scissor
        RenderUtil.startGlScissor((scaledResolution.getScaledWidth() / 2f) - (getWidth() / 2f), (scaledResolution.getScaledHeight() / 2f) - (getHeight() / 2f) + 14.5, getWidth(), getHeight() - 26.5f);

        float lineY = (scaledResolution.getScaledHeight() / 2f) - (getHeight() / 2f) + getHeight() - 26;

        Collections.reverse(lines);

        for (String string : lines) {
            renderText(string, (scaledResolution.getScaledWidth() / 2f) - (getWidth() / 2f) + 2, lineY, -1);
            lineY -= 11;
        }

        Collections.reverse(lines);

        // End scissor
        RenderUtil.endGlScissor();

        guiTextField.drawTextBox();
    }

    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {
        guiTextField.mouseClicked(mouseX, mouseY, mouseButton);
    }

    public void keyTyped(char typedChar, int keyCode) {
        if (keyCode == Keyboard.KEY_RETURN) {
            Paragon.INSTANCE.getCommandManager().handleCommands(guiTextField.getText(), true);

            guiTextField.setText("");
            guiTextField.setFocused(false);
            return;
        }

        guiTextField.textboxKeyTyped(typedChar, keyCode);
    }

    public void addLine(String line) {
        lines.add(line);
    }

    /**
     * Gets the title
     *
     * @return The title
     */
    public String getTitle() {
        return title;
    }

    /**
     * Gets the width
     *
     * @return The width
     */
    public float getWidth() {
        return width;
    }

    /**
     * Gets the height
     *
     * @return The height
     */
    public float getHeight() {
        return height;
    }
}

File: src/main/kotlin/com/paragon/impl/ui/console/Console.kt
package com.paragon.impl.ui.console

import com.paragon.Paragon
import com.paragon.util.render.font.FontUtil
import com.paragon.impl.module.client.Colours
import com.paragon.util.Wrapper
import com.paragon.util.render.RenderUtil
import net.minecraft.client.gui.GuiTextField
import net.minecraft.client.gui.ScaledResolution
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import org.lwjgl.input.Keyboard
import java.awt.Color

@SideOnly(Side.CLIENT)
class Console(private val title: String, val width: Float, val height: Float) : Wrapper {

    // List of lines
    private val lines: MutableList<String> = ArrayList(5)
    private var guiTextField: GuiTextField

    init {
        val scaledResolution = ScaledResolution(minecraft)
        guiTextField = GuiTextField(
            0, minecraft.fontRenderer, (scaledResolution.scaledWidth / 2 - width / 2).toInt() + 7, (scaledResolution.scaledHeight / 2 - height / 2 + height - 17).toInt(), width.toInt() - 14, 11
        )
    }

    fun init() {
        val scaledResolution = ScaledResolution(minecraft)
        guiTextField = GuiTextField(
            0, minecraft.fontRenderer, (scaledResolution.scaledWidth / 2 - width / 2).toInt() + 7, (scaledResolution.scaledHeight / 2 - height / 2 + height - 17).toInt(), width.toInt() - 14, 11
        )
    }

    fun draw() {
        val scaledResolution = ScaledResolution(minecraft)

        val x = (scaledResolution.scaledWidth / 2f) - (width / 2f)
        val y = (scaledResolution.scaledHeight / 2f) - (height / 2f)

        RenderUtil.drawRoundedRect(x, y, width, height, 10f, Color(20, 20, 25))
        RenderUtil.drawRoundedRect(x, y + 17.5f, width, 2f, 1f, Colours.mainColour.value)
        RenderUtil.drawRoundedOutline(x, y, width, height, 10f, 2f, Colours.mainColour.value)

        FontUtil.drawStringWithShadow(title, x + 7f, y + 7f, Color.WHITE)

        lines.reverse()

        RenderUtil.pushScissor(
            (scaledResolution.scaledWidth / 2f) - (width / 2f), (scaledResolution.scaledHeight / 2f) - (height / 2f) + 20f, width, (height - 26.5f)
        )

        var lineY = (scaledResolution.scaledHeight / 2f) - (height / 2f) + height - 30

        for (string in lines) {
            FontUtil.drawStringWithShadow(string, (scaledResolution.scaledWidth / 2f) - (width / 2f) + 5, lineY, Color.WHITE)
            lineY -= 11
        }

        lines.reverse()

        RenderUtil.popScissor()

        if (!guiTextField.text.startsWith("> ")) {
            guiTextField.text = "> " + guiTextField.text
        }

        guiTextField.drawTextBox()
    }

    fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        guiTextField.mouseClicked(mouseX, mouseY, mouseButton)
    }

    fun keyTyped(typedChar: Char, keyCode: Int) {
        if (keyCode == Keyboard.KEY_RETURN) {
            Paragon.INSTANCE.commandManager.handleCommands(guiTextField.text.substring(2), true)
            guiTextField.text = ""
            guiTextField.isFocused = false
            return
        }
        guiTextField.textboxKeyTyped(typedChar, keyCode)
    }

    fun addLine(line: String) {
        lines.add(line)
    }

}